### ISSUE TITLE
defined(__COUNTER__) results in an unexpected extension diagnostic

### ISSUE THREAD

**Issue by rnichollx** [llvm/llvm-project#196474]
```
/home/rpnx/quxlang/dev-workspace/sources/benchmark/include/benchmark/benchmark.h:1483:30: error: '__COUNTER__' is a C2y extension [-Werror,-Wc2y-extensions]
 1483 | #if defined(__COUNTER__) && (__COUNTER__ + 1 == __COUNTER__ + 0)
```

Clang should not error when `__COUNTER__` is disabled by compiler flags, the defined clause should return false and then clang should continue on without an error.

**Comment by AaronBallman** [comment]
`__COUNTER__` isn't disabled by any compiler flags; it's either a standardized preprocessor feature or an extension. It would be deeply odd for `-Werror` to change the behavior of how `defined()` resolves. I believe this is behaving by design (also, benchmark addressed the deficiency upstream: https://github.com/google/benchmark/commit/ff773f8c97717424a7855ce0933b036b79628a02)

**Comment by rnichollx** [comment]
> `__COUNTER__` isn't disabled by any compiler flags; it's either a standardized preprocessor feature or an extension. It would be deeply odd for `-Werror` to change the behavior of how `defined()` resolves. I believe this is behaving by design (also, benchmark addressed the deficiency upstream: [google/benchmark@ff773f8](https://github.com/google/benchmark/commit/ff773f8c97717424a7855ce0933b036b79628a02))

The problem is that I didn't set `-Werror`, even after disabling `-Werror`, this behavior persists. I have to pass `-Wno-c2y-extensions` or similar to prevent the error. When I disable extensions, that _shouldn't_ make them errors in ifdef statements.

The entire purpose of checking `#if defined(__COUNTER__)` is to check if `__COUNTER__` is available to use, if it's not, then this should not be defined at all.

It seems this `-Wc2y-extensions` is enabled by default or another flag in a non-intuitive way, possibly by setting the C++ standard to a specific value and disabling compiler extensions. I didn't manually add `-Werror` or `-Wc2y-extensions` and disabled them from the benchmark code, and I still got this error.

Regardless, if c2y-extensions are disabled, then `__COUNTER__` shouldn't be defined at all, instead of being a error.

**Comment by AaronBallman** [comment]
> The problem is that I didn't set `-Werror`, even after disabling `-Werror`, this behavior persists. I have to pass `-Wno-c2y-extensions` or similar to prevent the error. When I disable extensions, that _shouldn't_ make them errors in ifdef statements.

Are you passing `-pedantic-errors`?

> It seems this -Wc2y-extensions is enabled by default or another flag in a non-intuitive way,

It's enabled via `-pedantic` and `-pedantic-errors` because those turn on all extension warnings (and the latter treats them as errors).

> Regardless, if c2y-extensions are disabled, then __COUNTER__ shouldn't be defined at all, instead of being a error.

That is not a change we will make. Warning flags should not modify the behavior of code (beyond `-Werror` failing translation). We have experience with this via `-Wwrite-strings` and we don't want to repeat that mistake.

That said, we are expecting to suppress the diagnostic when expanded from a system macro which should improve this situation.
